src/main/scala/com/hortonworks/spark/sql/hive/llap/HS2JDBCWrapper.scala

```diff
@@ -166,12 +168,38 @@ class JDBCWrapper {
   def getConnector(
       userProvidedDriverClass: Option[String],
       url: String,
-      userName: String): Connection = {
+      userName: String,
+      dbcp2Configs: String): Connection = {
     log.debug(s"${userProvidedDriverClass.getOrElse("")} $url $userName password")
     val subprotocol = new URI(url.stripPrefix("jdbc:")).getScheme
     val driverClass: Class[Driver] = getDriverClass(subprotocol, userProvidedDriverClass)
-    registerDriver(driverClass.getCanonicalName)
-    DriverManager.getConnection(url, userName, "password")
+
+    connectionPools.get(userName) match {
+      case Some(d) =>
+        d.getConnection
+      case None =>
+        val datasource = new BasicDataSource
+        datasource.setDriverClassName(driverClass.getCanonicalName)
+        datasource.setUrl(url)
+        datasource.setUsername(userName)
+        // for hdp older version support without dbcp2 configurations
+        if(dbcp2Configs == null) {
+            datasource.setInitialSize(2)
+            datasource.setMaxConnLifetimeMillis(2000)
+            datasource.setMaxTotal(100)
+            datasource.setMaxIdle(50)
```

Comment:
when is connection considered idle ? 
Keeping 50 idle connection is wasting resources in HS2. Keep max of 1 idle connections around ?
If the connections are actively used, I don't think they would get marked as idle.